### PR TITLE
Add webdriver tests for dev and prod

### PR DESCRIPTION
**Pull request by maggie-lou**
---

**Version bump**: TODO

### CODE REVIEW

**Comment by maggie-lou** [comment]
Current dependencies on/for this PR:
* master
  * **PR #3260** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3260" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3260?utm_source=stack-comment).

File: enterprise/server/test/webdriver/invocation/invocation_test.go [hunk 1]
```diff
@@ -2,18 +2,90 @@ package invocation_test
...
 	"github.com/buildbuddy-io/buildbuddy/server/testutil/testbazel"
 	"github.com/buildbuddy-io/buildbuddy/server/testutil/webtester"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 )
 
-func TestAuthenticatedInvocation_LogUploadEnabled(t *testing.T) {
+var (
+	targetStr   = flag.String("webdriver_target", "local", "Target that should be tested by the webdriver.")
+	devSSOSlug  = flag.String("webdriver_dev_slug", "", "Dev SSO slug to login.")
```
Comment (maggie-lou) [inline enterprise/server/test/webdriver/invocation/invocation_test.go:18]: The plan is to save these SSO slugs as secrets in the github action that will do the release. Then after we deploy dev for example, we can run this webdriver test on the dev target with the dev SSO secret

File: app/invocation/cache_requests_card.tsx [hunk 2]
```diff
@@ -302,7 +302,7 @@ export default class CacheRequestsCardComponent extends React.Component<CacheReq
           {/* Filtering controls */}
           <div className="separator" />
           <label>Show</label>
-          <Select value={this.getFilterIndex()} onChange={this.onChangeFilter.bind(this)}>
+          <Select id="filter-cache-requests" value={this.getFilterIndex()} onChange={this.onChangeFilter.bind(this)}>
```
Comment (bduffany) [inline app/invocation/cache_requests_card.tsx:305]: nit: we usually use `className` instead of `id`

File: enterprise/server/test/webdriver/invocation/invocation_test.go (hunk 1 above)
Comment (bduffany) [inline enterprise/server/test/webdriver/invocation/invocation_test.go:18]: Instead of `local`, `dev`, `prod`, wdyt about having `local` and `remote`, where remote has to be explicitly configured with `--remote_app_endpoint` and `--remote_sso_slug`? I think it might be cleaner since this binary would only have to know about one k8s env at a time, instead of having the potential to accidentally communicate with multiple envs

File: enterprise/server/test/webdriver/invocation/invocation_test.go [hunk 3]
```diff
@@ -2,18 +2,90 @@ package invocation_test
...
+	prodSSOSlug = flag.String("webdriver_prod_slug", "", "Prod SSO slug to login.")
+)
+
+type webdriverTarget interface {
```
Comment (bduffany) [inline enterprise/server/test/webdriver/invocation/invocation_test.go:22]: Can we move these structs and flags to a shared package, maybe `server/testutil/testtarget` or `server/testutil/webtest_target` ? (doesn't need to be under enterprise/, since we might want to have some library functions which depend on the Target, but can be used with OSS webdriver tests). That way when we add more webdriver tests we don't need need to duplicate these

File: app/invocation/cache_requests_card.tsx (hunk 2 above)
Comment (siggisim) [reply app/invocation/cache_requests_card.tsx:305]: Trying webdriver tests to classNames can be brittle (i.e. you could accidentally break a webdriver test by changing the style of an element).

A few options (I'm open to any of theses):
- We could have a special classname prefix that we know will be selected by webdriver, so we shouldn't change it
- We could use another attribute that's only used for webdriver ids like `debug-id="foo"`
- We could make IDs be our 
- We could roll with class names, and just be careful to search the webdriver tests before changing them

Here's a post on selenium locator best practices: https://www.screenster.io/selenium-locators-best-practices-7-helpful-tips-for-ui-testers/

Comment (bduffany) [reply app/invocation/cache_requests_card.tsx:305]: I like the `debug-id` idea; we might have to call it `data-debug-id` to make React/TypeScript happy.

File: server/testutil/webtester/webtester.go [hunk 4]
```diff
@@ -168,14 +182,22 @@ func HasClass(el *Element, class string) bool {
 // BuildBuddy-specific functionality
 // ===
 
-// Login uses the Web app to log into BuildBuddy as the default self-auth user.
-// It expects that no user is currently logged in, and that self-auth is
-// enabled.
-func Login(wt *WebTester, appBaseURL string) {
+// LoginSelfAuth uses the Web app to log into BuildBuddy as the default self-auth user.
+// It expects that no user is currently logged in, and that self-auth is enabled on the server.
+func LoginSelfAuth(wt *WebTester, appBaseURL string) {
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply server/testutil/webtester/webtester.go]: Done. Makes sense to have the local webdriver act as similarly to the remote ones as possible too

File: server/testutil/webtester/webtester.go [hunk 5]
```diff
@@ -87,6 +87,14 @@ func (wt *WebTester) FindAll(cssSelector string) []*Element {
 	return out
 }
 
+// FindName returns the element matching the given CSS name. Exactly one
+// element must be matched, otherwise the test fails.
+func (wt *WebTester) FindName(name string) *Element {
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply server/testutil/webtester/webtester.go]: Got rid of this

File: enterprise/server/test/webdriver/invocation/invocation_test.go [hunk 6]
```diff
@@ -2,18 +2,90 @@ package invocation_test
...
+		break
+	case "dev":
+		target = &devTarget{}
+		break
+	case "prod":
+		target = &prodTarget{}
+		break
+	default:
+		require.FailNowf(t, "invalid target", "%s is an invalid target for the webdriver invocation tests.", targetStr)
+	}
```
[the remark this answers is not in the document]
Comment (maggie-lou) [reply enterprise/server/test/webdriver/invocation/invocation_test.go]: Done

File: enterprise/server/test/webdriver/invocation/invocation_test.go (hunk 3 above)
Comment (maggie-lou) [reply enterprise/server/test/webdriver/invocation/invocation_test.go]: Moved it under /enterprise because OSS doesn't support auth. If we end up needing these drivers in the OSS tests, we can refactor this later to move the interface to OSS, but keep drivers with auth in /enterprise

File: enterprise/server/test/webdriver/invocation/invocation_test.go (hunk 1 above)
Comment (maggie-lou) [reply enterprise/server/test/webdriver/invocation/invocation_test.go]: Done

File: app/invocation/cache_requests_card.tsx (hunk 2 above)
Comment (maggie-lou) [reply app/invocation/cache_requests_card.tsx]: Good idea. Modified to use `debug-id`

**Review by siggisim — APPROVED** [review]

**Review by bduffany — APPROVED** [review]
